internal/cloudsql/instance.go

```diff
@@ -0,0 +1,192 @@
+// only be intialalized by the Instance struct as part of a refresh cycle.
+type refreshResult struct {
+	md     metadata
+	tlsCfg *tls.Config
+	err    error
+
+	// timer that triggers refresh, can be used to cancel.
+	timer *time.Timer
+	// indicates the struct is ready to read from
+	ready chan struct{}
+}
+
+// Cancel prevents the instanceInfo from starting, if it hasn't already started. Returns true if timer
+// was stopped successfully, or false if it has already started.
+func (i *refreshResult) Cancel() bool {
+	return i.timer.Stop()
+}
+
+// Wait blocks until the refreshResult attempt is completed.
+func (i *refreshResult) Wait(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-i.ready:
+		return i.err
+	}
+}
+
+// Instance manages the information used to connect to the Cloud SQL instance by periodically calling
+// the Cloud SQL Admin API. It automatically refreshes the required information every 55 minutes.
+type Instance struct {
+	connName
+	client *sqladmin.Service
+	key    *rsa.PrivateKey
+
+	resultGuard *sync.RWMutex
+	cur         *refreshResult
+	next        *refreshResult
+
+	// TODO: add a way to close
+}
+
+// NewInstance initializes a new Instance given an instance conneciton name
+func NewInstance(instanceConnName string, client *sqladmin.Service, key *rsa.PrivateKey) (*Instance, error) {
+	cn, err := parseConnName(instanceConnName)
+	if err != nil {
+		return nil, err
+	}
+	i := &Instance{cn, client, key, &sync.RWMutex{}, nil, nil}
+	// Kick off the inital refresh op asynchronously
+	i.resultGuard.Lock()
+	i.cur = i.scheduleRefresh(0)
+	i.next = i.cur
+	i.resultGuard.Unlock()
+	return i, nil
+}
+
+// Connect returns a secure, authorized net.Conn to a Cloud SQL instance.
+func (im *Instance) Connect(ctx context.Context) (net.Conn, error) {
+	im.resultGuard.RLock()
+	res := im.cur
+	im.resultGuard.RUnlock()
+	err := res.Wait(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Add better ipType support, including an opt to specify.
+	addr := net.JoinHostPort(res.md.ipAddrs["PUBLIC"], "3307")
+	conn, err := proxy.Dial(ctx, "tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	if c, ok := conn.(*net.TCPConn); ok {
+		if err := c.SetKeepAlive(true); err != nil {
+			return nil, fmt.Errorf("failed to set keep-alive: %w", err)
+		}
+		if err := c.SetKeepAlivePeriod(30 * time.Second); err != nil {
+			return nil, fmt.Errorf("failed to set keep-alive period: %w", err)
+		}
+	}
+
+	tlsConn := tls.Client(conn, res.tlsCfg)
+	if err := tlsConn.Handshake(); err != nil {
+		tlsConn.Close()
+		return nil, fmt.Errorf("handshake failed: %w", err)
+	}
+	return tlsConn, err
+}
+
+// scheduleRefresh schedules a refresh operation to be triggered after a given duration. The returned resultRefresh
+// can be used to either Cancel or Wait for the operations result.
+func (im *Instance) scheduleRefresh(d time.Duration) *refreshResult {
+	res := &refreshResult{}
+	res.ready = make(chan struct{})
```

Comment:
So I considered this originally, but feel a bit iffy on it since `timer` also needs to be set for this to be useful. I'm not sure it makes sense to split the initialization for `ready` and `timer` into different locations. In order to put `timer` into `newRefresh`, I'd have to pass the instance (and mutate it), which seems worse than just keeping it in one place here. WDYT?

Reply:
That makes sense.